neqo-common/src/qlog.rs

```diff
@@ -144,3 +144,71 @@ pub fn new_trace(role: Role) -> qlog::TraceSeq {
         }),
     }
 }
+
+#[cfg(test)]
+mod test {
+    use super::{new_trace, NeqoQlog};
+    use crate::Role;
+    use qlog::{
+        events::{Event, EventImportance},
+        streamer::QlogStreamer,
+    };
+    use std::io::Cursor;
+
+    const EV_DATA: qlog::events::EventData =
+        qlog::events::EventData::SpinBitUpdated(qlog::events::connectivity::SpinBitUpdated {
+            state: true,
+        });
+
+    const EXPECTED_LOG_HEADER: &str = "\u{1e}{\"qlog_version\":\"0.3\",\"qlog_format\":\"JSON-SEQ\",\"trace\":{\"vantage_point\":{\"name\":\"neqo-Client\",\"type\":\"client\"},\"title\":\"neqo-Client trace\",\"description\":\"Example qlog trace description\",\"configuration\":{\"time_offset\":0.0},\"common_fields\":{\"reference_time\":0.0,\"time_format\":\"relative\"}}}\n";
+    const EXPECTED_LOG_EVENT: &str = "\u{1e}{\"time\":0.0,\"name\":\"connectivity:spin_bit_updated\",\"data\":{\"state\":true}}\n";
+
+    fn new_neqo_qlog() -> NeqoQlog {
+        let mut trace = new_trace(Role::Client);
+        // Set reference time to 0.0 for testing.
+        trace.common_fields.as_mut().unwrap().reference_time = Some(0.0);
+
+        let streamer = QlogStreamer::new(
+            qlog::QLOG_VERSION.to_string(),
+            None,
+            None,
+            None,
+            std::time::Instant::now(),
+            trace,
+            EventImportance::Base,
+            Box::new(Cursor::new(Vec::new())),
+        );
+        let log = NeqoQlog::enabled(streamer, "");
+        assert!(log.is_ok());
+        log.unwrap()
+    }
+
+    fn log_contents(log: &NeqoQlog) -> String {
+        // TODO: Figure out a way to make this less ugly.
+        #[allow(clippy::borrowed_box)]
+        let w: &Box<std::io::Cursor<Vec<u8>>> = unsafe {
+            #[allow(clippy::transmute_ptr_to_ptr)]
+            std::mem::transmute(log.inner.borrow_mut().as_mut().unwrap().streamer.writer())
```

Comment:
Please try to avoid using `unsafe`.  You should be able to do that by implementing Writer on a struct that holds a shared reference to the underlying Vec.  You'll need to use Arc for the shared reference (stupid Send+Sync requirement in the qlog interface).

Reply:
I'll try to work it out. That particular piece of magic came from GitHub Copilot, because my Rust-fu isn't there yet to understand the intricacies here.

Reply:
This might help...

```rust
#[derive(Clone)]
struct SharedVec {
  buf: Arc<Mutex<Cursor<Vec<u8>>>>,
}
impl Writer for SharedVec {
    fn write(&mut self, buf: &[u8]) -> Result<usize> {
        self.buf.lock().write(buf)
    }
    fn flush(&mut self) -> Result<()> {
        self.buf.lock().flush(buf)
    }
}
```

Reply:
OK, I am creating streamer as 
```rust
    let streamer = QlogStreamer::new(
        qlog::QLOG_VERSION.to_string(),
        None,
        None,
        None,
        std::time::Instant::now(),
        trace,
        EventImportance::Base,
        Box::new(SharedVec {
            buf: Arc::new(Mutex::new(Cursor::new(Vec::new()))),
        }),
    );
```
but am then struggling to get at the `SharedVec`:
```rust
pub fn neqo_qlog_contents(log: &NeqoQlog) -> String {
    let w = log
        .inner()
        .borrow_mut()
        .as_mut()
        .unwrap()
        .streamer()
        .writer();

    String::from_utf8(w.as_ref().get_ref().clone()).unwrap() // Not like this?
}
```

Reply:
Create it like this instead:

```rust
let buf = Arc::new(Mutex::new(Cursor::new(Vec::new())));
let other_buf = Arc::clone(&buf);
let streamer = ... SharedVec { buf } ...;
```

You will then have to return other_buf, which you can query the contents of at will.